src/sage/stats/distributions/discrete_gaussian_lattice.py

```diff
@@ -279,34 +451,102 @@ def __init__(self, B, sigma=1, c=None, precision=None):
 
             sage: v = vector(ZZ, n, (-3, -3))
             sage: v.set_immutable()
-            sage: while v not in counter:
-            ....:     add_samples(1000)
-            sage: while abs(m*f(v)*1.0/c/counter[v] - 1.0) >= 0.1:                      # needs sage.symbolic
+            sage: while v not in counter: add_samples(1000)
+            sage: while abs(m*f(v)*1.0/nf/counter[v] - 1.0) >= 0.1:                     # needs sage.symbolic
             ....:     add_samples(1000)
 
+            sage: counter = defaultdict(Integer)
             sage: v = vector(ZZ, n, (0, 0))
             sage: v.set_immutable()
             sage: while v not in counter:
             ....:     add_samples(1000)
-            sage: while abs(m*f(v)*1.0/c/counter[v] - 1.0) >= 0.1:                      # needs sage.symbolic
+            sage: while abs(m*f(v)*1.0/nf/counter[v] - 1.0) >= 0.1:                     # needs sage.symbolic
             ....:     add_samples(1000)
 
-            sage: from sage.stats.distributions.discrete_gaussian_lattice import DiscreteGaussianDistributionLatticeSampler
-            sage: qf = QuadraticForm(matrix(3, [2, 1, 1,  1, 2, 1,  1, 1, 2]))
-            sage: D = DiscreteGaussianDistributionLatticeSampler(qf, 3.0); D            # needs sage.symbolic
-            Discrete Gaussian sampler with σ = 3.000000, c=(0, 0, 0) over lattice with basis
+        Spherical covariance are automatically handled::
+
+            sage: distributions.DiscreteGaussianDistributionLatticeSampler(ZZ^3, sigma=Matrix(3, 3, 2))
+            Discrete Gaussian sampler with Gaussian parameter σ = 2.00000000000000, c=(0, 0, 0) over lattice with basis
+            <BLANKLINE>
+            [1 0 0]
+            [0 1 0]
+            [0 0 1]
+
+        The sampler supports non-spherical covariance in the form of a Gram
+        matrix::
+
+            sage: n = 3
+            sage: Sigma = Matrix(ZZ, [[5, -2, 4], [-2, 10, -5], [4, -5, 5]])
+            sage: c = vector(ZZ, [7, 2, 5])
+            sage: D = distributions.DiscreteGaussianDistributionLatticeSampler(ZZ^n, Sigma, c)
+            sage: nf = D._normalisation_factor_zz(); nf # This has not been properly implemented
+            63.76927...
+            sage: while v not in counter: add_samples(1000)
+            sage: while abs(m*f(v)*1.0/nf/counter[v] - 1.0) >= 0.1: add_samples(1000)
```

Comment:
This test seems wrong... 

Unless we are lucky, this will never finish. Note that `counter` was not reset, we are running the test with the vector from the previous sample, and sometimes the test passes even with a different normalization factor. But if the test doesn't pass on the first few attempts, it won't pass because `m` tends to infinity while `counter[v]` remains constant.

Make sure you reset `counter` and you will see this test looping forever always.

Reply:
Sorry about this, I didn’t test with —long much (but why did CI not catch it?)

Can you fix it for me? I am in exam seasons right now, sorry

Reply:
Ping @grhkm21 @tscrim

Can we either get this fixed before 10.4 or revert it until it can be fixed properly?

Reply:
Indeed, the test is bad because it does not reset `v` nor `counter`, so it is using the same one from the previous test (which has a similar enough normalization factor). This is why it wasn't getting picked up generally by the tester.

I don't think I would be able to fix this myself. I propose we effectively disable this test with `# known bug`. Reverting the whole PR feels a bit extreme since the normalization doesn't seem to be a key part of the PR (it is hidden, not called from anywhere as far as I can tell, and emits a warning).

Reply:
Sorry for the late reply. I fixed it, it's just copying some lines above. Please kindly review #38209 and sorry for the bad code :(